### PR TITLE
Change to koa-based API

### PR DESCRIPTION
**Pull request by justinvdm**
# why not klein

The plan is for Numi's API to validate requests, then delegate to internal APIs (with some glue code where necessary if some of those internal APIs are not Numi-specific).

I initially planned to use klein for a few reasons:
- some people on the team are familiar with it
- it is lightweight
- it is async

After getting started on request validation and processing, I found myself wanting to be able to define pipelines for processing requests like this:

```
creating:
  - read body as json
  - validate against schema
  - apply defaults given in schema
  - process request
  - serialize response as json

reading:
  - validate query params
  - process request
  - serialize response as json

updating and patching:
  - read body as json
  - ignore read only fields in schema
  - validate against schema
  - apply defaults given in schema
  - serialize response as json

remove:
  - validate query params
  - process request
  - serialize response as json
```

There are definitely ways of going about doing those things with klein, but there are frameworks out there with middleware features that make processing requests in that way easier to define, reason about and maintain.

I decided to consider two node.js frameworks, [express](https://github.com/expressjs/express) and [koa](https://github.com/koajs/koa/tree/v2.x) (express's successor): 
- these frameworks match the way I've been thinking of doing request validation and processing via middleware
- people on the team are already familiar with js, and these frameworks are small enough that a learning curve isn't a concern
- these frameworks are lightweight
- node.js, so async
# why koa over express

Koa has seen less production than express, but I would be surprised if there are more people out there using klein in production than koa. I was initially leaning on using express, given it has seen the most production, but decided to go with koa for mainly these reasons:
- it is built with promises in mind (especially helpful considering how this API is to be delegating to other APIs)
- it is easier to reason about its approach to error handling
- it allows for more control over the order in which a request is processed by each middleware
# why no babel

es7's `async` and `wait` compliment Koa's design, and many people seem to be using transpiling in order to use future language features in node.js. I've decided against this:
- a build step is an unnecessary complication for our case
- node v6 has the most important es features we need
- debugging transpiled code is a nightmare

### CODE REVIEW

**Comment by justinvdm** [comment]
@rudigiesler read for review

**Comment by justinvdm** [comment]
Made some changes so that processing of request data against json schemas is done in the programmatic api instead of in middleware.

_Update_: reverted this change, decided it is better to do validation and other processing against a schema in middleware instead.

_Update_: changed request parsing again, this time changing schema processing components to functions instead of middlewares and changing the method middlewares (the middlewares that wrap route handler functions to process the request against a schema) to handle calling those functions.

Ready for review.

File: src/api/index.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+const users = require('src/api/users');
+
+
+module.exports = {
+  users
```
Comment (hodgestar) [inline src/api/index.js]: Should this be `users: users`?

File: src/middleware/api/methods.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+
+function update(schema, fn) {
+  return method((ctx, args, next) => Promise.resolve(ctx.request.body)
+    .then(d => omitReadOnly(schema, d))
+    .then(effect(d => validate(schema, d)))
+    .then(d => defaults(schema, d))
+    .then(d => fn(...args, d))
+    .then(res => { ctx.body = res; })
```
Comment (hodgestar) [inline src/middleware/api/methods.js:34]: How will we set different HTTP status codes? E.g. 200 vs 201?

File: src/routes/index.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+const users = require('src/routes/users');
+
+
+module.exports = []
+  .concat(users);
```
Comment (hodgestar) [inline src/routes/index.js]: Could we not just use `module.exports = [ users ];`?

File: tests/middlewares/api/methods.test.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+          a: 2,
+          b: 3,
+          d: {foo: 23}
+        })
+        .end(done);
+    });
+  });
+
+  describe('remove', () => {
+    it('should use the apy call result as the response body', done => {
```
Comment (hodgestar) [inline tests/middlewares/api/methods.test.js]: s/apy/api/

**Comment by hodgestar** [comment]
Left two small questions, otherwise looks good!

File: src/middleware/api/methods.js (hunk 2 above)
Comment (justinvdm) [inline src/middleware/api/methods.js:34]: Thanks for catching this! I think only `create` methods will need to set the response code to `201`, will update `create()` to do this.

**Comment by justinvdm** [comment]
Ready for re-review.

**Comment by hodgestar** [comment]
:+1:

**Pull request merged**